crates/composition/src/lib.rs

```diff
@@ -2,33 +2,46 @@
 
 //! GraphQL schema composition.
 
-mod compose_supergraph;
-mod context;
+mod compose;
+mod composition_ir;
 mod diagnostics;
+mod emit_federated_graph;
 mod ingest_subgraph;
 mod result;
 mod strings;
 mod subgraphs;
-mod supergraph;
 
 pub use self::{diagnostics::Diagnostics, result::CompositionResult, subgraphs::Subgraphs};
 
-use self::{context::Context, strings::StringId, supergraph::Supergraph};
+use self::{
+    compose::{compose_subgraphs, ComposeContext},
+    emit_federated_graph::emit_federated_graph,
+};
 
-/// Compose subgraphs into a supergraph
+/// Compose subgraphs into a federated graph.
 pub fn compose(subgraphs: &Subgraphs) -> CompositionResult {
-    let mut context = Context {
-        subgraphs,
-        supergraph: Supergraph::default(),
-        diagnostics: Diagnostics::default(),
-    };
+    let mut diagnostics = Diagnostics::default();
+    let mut context = ComposeContext::new(subgraphs, &mut diagnostics);
 
-    compose_supergraph::build_supergraph(&mut context);
+    compose_subgraphs(&mut context);
 
-    let supergraph_sdl = context.supergraph.render(&context.subgraphs.strings);
+    if context.diagnostics.any_fatal() {
+        return CompositionResult {
+            diagnostics,
+            federated_graph: Default::default(),
+        };
+    }
+
+    let federated_graph = emit_federated_graph(context.into_ir(), subgraphs);
 
     CompositionResult {
-        supergraph_sdl,
-        diagnostics: context.diagnostics,
+        federated_graph,
+        diagnostics,
     }
 }
+
+fn push_and_return_id<T, Id>(elems: &mut Vec<T>, new_elem: T, make_id: fn(usize) -> Id) -> Id {
+    let id = make_id(elems.len());
+    elems.push(new_elem);
+    id
+}
```

Comment:
Kinda wondering about this function as well.  Maybe just me, but feels like the make_id parameter is obscuring what's going on without providing much benefit?

You _could_ just return a `usize` and do `federated::FieldTypeId(push_and_return_id(&mut whatever, Whatever))`?

Dunno, don't have strong feelngs, just a thought.

Reply:
I might personally have been tempted to stick this on a `Vec` extension trait as well, to save the  `&mut`.

`FieldTypeId(self.field_types.push_and_return_id(FieldType { .. }))` seems like it might be alright?  Dunno, definitely not an important comment

Reply:
Yes both good ideas. I'll experiment and see what look and feels better.

Reply:
I went with the extension trait, I think it does read better.